### ISSUE TITLE
Compilation issues

### ISSUE THREAD

**Issue by usmanmehmoodSSP** [higaski/EmbeddedCMake#1]
Following your tutorial, and doing some debugging along the way, I used this command to configure the CMake project
```
cmake -G "Unix Makefiles" -Bbuild -DCMAKE_TOOLCHAIN_FILE=./cmake/arm_none_eabi_gcc.cmake -DCMAKE_BUILD_TYPE=Debug .
```

which gives me this output
```
CMake Warning:
  Ignoring extra path from command line:

   "C:/Users/my_username/Desktop/EmbeddedCMake/cmake/arm_none_eabi_gcc.cmake"


-- The ASM compiler identification is GNU
-- Found assembler: C:/msys64/mingw64/bin/cc.exe
-- The C compiler identification is GNU 12.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/msys64/mingw64/bin/cc.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done (0.6s)
-- Generating done (0.0s)
-- Build files have been written to: C:/Users/my_username/Desktop/EmbeddedCMake/build
```

After which I use `make -C build`, and among other warnings, it gives me this error
```
C:\Users\U2E2C~1.MEH\AppData\Local\Temp\cckiaYF5.s: Assembler messages:
C:\Users\U2E2C~1.MEH\AppData\Local\Temp\cckiaYF5.s:344: Error: no such instruction: `cpsid i'
make[2]: *** [CMakeFiles/EmbeddedCMake.dir/Core/Src/main.c.obj] Error 1
make[2]: Leaving directory `C:/Users/my_username/Desktop/EmbeddedCMake/build'
make[1]: *** [CMakeFiles/EmbeddedCMake.dir/all] Error 2
make[1]: Leaving directory `C:/Users/my_username/Desktop/EmbeddedCMake/build'
make: *** [all] Error 2
make: Leaving directory `C:/Users/my_username/Desktop/EmbeddedCMake/build'
```

Can I have any insight as to why this is happening?

**Comment by higaski** [comment]
Hello

It looks like CMake is ignoring the path you've provided to your toolchain file

```sh
  Ignoring extra path from command line:

   "C:/Users/my_username/Desktop/EmbeddedCMake/cmake/arm_none_eabi_gcc.cmake"
```

CMake then continuous to pick the next best C compiler it can find which seems to be Mingw64 GCC 12.2.0. This is your systems toolchain for x86 though and not the ARM toolchain you've intended to use.  The error message you get when running make

```sh
C:\Users\U2E2C~1.MEH\AppData\Local\Temp\cckiaYF5.s:344: Error: no such instruction: `cpsid i'
```

is the compiler complaining that there is no CPSID instruction on x86.


Sadly I can't tell why CMake doesn't like your path. Maybe it's a backslash windows thing...? Are there any whitespaces in the original path?

**Comment by usmanmehmoodSSP** [comment]
Thanks for the quick response :)

I was suspecting something similar. 
So I manually set the compiler paths in the main CMake

```bash
set(CMAKE_ASM_COMPILER "C:\\emb_tools\\arm-gnu-toolchain-12.2.rel1-mingw-w64-i686-arm-none-eabi\\bin\\arm-none-eabi-as.exe")
set(CMAKE_C_COMPILER "C:\\emb_tools\\arm-gnu-toolchain-12.2.rel1-mingw-w64-i686-arm-none-eabi\\bin\\arm-none-eabi-gcc.exe")
```

And it still gives the same error, albeit in a slightly different way. (I don't know if I'd call it progress)

```bash
PS C:\Users\u.mehmood\Desktop\EmbeddedCMake> make -C build
make: Entering directory `C:/Users/u.mehmood/Desktop/EmbeddedCMake/build'
make[1]: Entering directory `C:/Users/u.mehmood/Desktop/EmbeddedCMake/build'
make[2]: Entering directory `C:/Users/u.mehmood/Desktop/EmbeddedCMake/build'
make[2]: Leaving directory `C:/Users/u.mehmood/Desktop/EmbeddedCMake/build'
make[2]: Entering directory `C:/Users/u.mehmood/Desktop/EmbeddedCMake/build'
[  3%] Building C object CMakeFiles/EmbeddedCMake.dir/Core/Src/main.c.obj
C:\Users\U2E2C~1.MEH\AppData\Local\Temp\ccgxcVy9.s: Assembler messages:
C:\Users\U2E2C~1.MEH\AppData\Local\Temp\ccgxcVy9.s:463: Error: selected processor does not support `cpsid i' in ARM mode
make[2]: *** [CMakeFiles/EmbeddedCMake.dir/Core/Src/main.c.obj] Error 1    
make[2]: Leaving directory `C:/Users/u.mehmood/Desktop/EmbeddedCMake/build'
make[1]: *** [CMakeFiles/EmbeddedCMake.dir/all] Error 2
make[1]: Leaving directory `C:/Users/u.mehmood/Desktop/EmbeddedCMake/build'
make: *** [all] Error 2
make: Leaving directory `C:/Users/u.mehmood/Desktop/EmbeddedCMake/build'
```

I also checked my `PATH`, there's no spaces in it. There is however a dot in the user name (`u.mehmood`) but I don't know if it should matter. 

In any case, I will try to recreate the same in another system that ONLY has the `arm-none-eabi` compiler.

**Comment by higaski** [comment]
Seems like its picking the right compiler now but the architecture flags are wrong.

To keep the toolchain files free of hardcoded architecture flags I usually use a variable I set on the command line before passing the toolchain file to CMake.
```cmake
set(CMAKE_ASM_FLAGS "${ARCH}")
set(CMAKE_C_FLAGS "${ARCH}")
set(CMAKE_CXX_FLAGS "${ARCH}")
```

E.g. for using a Cortex M4 with floating point unit I'd invoke CMake like so
```sh
cmake -Bbuild -GNinja -DARCH="-mcpu=cortex-m4 -mfloat-abi=hard" -DCMAKE_TOOLCHAIN_FILE=arm_none_eabi_gc.cmake
```

**Comment by usmanmehmoodSSP** [comment]
The architecture flags definitely helped. 

I also noticed that the `arm_none_eabi_gcc.cmake` toolchain file was not being picked up from relative path, so when I gave it the absolute path, it all worked and compiled successfully. 

```bash
cmake -GNinja -Bbuild -DARCH="-mcpu=cortex-m4 -mfloat-abi=hard" -DCMAKE_TOOLCHAIN_FILE=C:\Users\u.mehmood\Desktop\EmbeddedCMake\cmake\arm_none_eabi_gcc.cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_MAKE_PROGRAM=ninja .
```

I did remove other variants of GCC from my system. But even after putting them back, it still works as it should so it was probably the toolchain file being skipped that was causing these issues. 

Anyway, thanks a bunch for the guidance and more importantly for your tutorial and repo. 😃 
It is super helpful and without it I would have been spiralling in CMake docs forever.
